src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,102 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private Object[] values;
+    private int size;
+    private int capacity = DEFAULT_CAPACITY;
+
+    public ArrayList() {
+        values = new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        resize();
+        values[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index out of bounds!");
+        }
+        resize();
+        System.arraycopy(values, index, values, index + 1, size - index);
+        values[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index out of bounds!");
+        }
+        return (T) values[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index out of bounds!");
+        }
+        values[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index out of bounds!");
+        }
+        Object removedValue = values[index];
+        System.arraycopy(values, index + 1, values, index, size - index - 1);
+        size--;
+        return (T) removedValue;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        T removedValue = null;
+        boolean isFind = false;
+        for (int i = 0; i < size; i++) {
+            if (values[i] != null && ((T) values[i]).equals(element) || values[i] == element) {
```

Comment:
```suggestion
            if (values[i] != null && (values[i]).equals(element) || values[i] == element) {
```
Let's try to remove casting of values[i], it should work without it